metrics/internal/collectors/storage-auto-scaler.go

```diff
@@ -0,0 +1,179 @@
+		s.StorageAutoScalerMetadata,
+		s.AutoScalerPhaseStatus,
+		s.StorageCapacityReached,
+	}
+	for _, d := range ds {
+		ch <- d
+	}
+}
+
+func (s *StorageAutoScalerCollector) Collect(ch chan<- prometheus.Metric) {
+	sasLister := NewStorageAutoScalerLister(s.Informer.GetIndexer())
+	storageAutoScalers, err := sasLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("couldn't list StorageAutoScaler elements: %v", err)
+		return
+	} else if len(storageAutoScalers) == 0 {
+		return
+	}
+	var wg sync.WaitGroup
+	wg.Add(3) // we are waiting for THREE go routines to finish
+	go s.collectStorageAutoScalerMetadata(ch, storageAutoScalers, &wg)
+	go s.collectStorageAutoScalerPhaseStatuses(ch, storageAutoScalers, &wg)
+	go s.collectStorageCapacityReachedStatuses(ch, storageAutoScalers, &wg)
+	wg.Wait()
+}
+
+func (s *StorageAutoScalerCollector) Run(stopCh <-chan struct{}) {
+	go s.Informer.Run(stopCh)
+}
+
+func (s *StorageAutoScalerCollector) collectStorageAutoScalerMetadata(ch chan<- prometheus.Metric, storageAutoScalers []*v1.StorageAutoScaler, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for _, storageAutoScaler := range storageAutoScalers {
+		if storageAutoScaler == nil {
+			continue
+		}
+		ch <- prometheus.MustNewConstMetric(
+			s.StorageAutoScalerMetadata,
+			prometheus.GaugeValue,
+			float64(1),
+			storageAutoScaler.Name,
+			storageAutoScaler.Namespace,
+			storageAutoScaler.Spec.StorageCluster.Name,
+			storageAutoScaler.Spec.DeviceClass,
+		)
+	}
+}
+
+func (s *StorageAutoScalerCollector) collectStorageAutoScalerPhaseStatuses(ch chan<- prometheus.Metric, storageAutoScalers []*v1.StorageAutoScaler, wg *sync.WaitGroup) {
+	defer wg.Done()
+	for _, storageAutoScaler := range storageAutoScalers {
+		if storageAutoScaler == nil {
+			continue
+		}
+		var phaseStatus = NotStartedPhaseStatus
+		switch storageAutoScaler.Status.Phase {
+		case v1.StorageAutoScalerPhaseFailed:
+			phaseStatus = FailedPhaseStatus
+		case v1.StorageAutoScalerPhaseSucceeded:
+			phaseStatus = succeededPhaseValidation(storageAutoScaler.Status.LastExpansion.StartTime, storageAutoScaler.Status.LastExpansion.CompletionTime)
+		case v1.StorageAutoScalerPhaseInProgress:
+			phaseStatus = InProgressPhaseStatus
+		default:
+			phaseStatus = NotStartedPhaseStatus
+		}
+		ch <- prometheus.MustNewConstMetric(
+			s.AutoScalerPhaseStatus,
+			prometheus.GaugeValue,
+			float64(phaseStatus),
+			storageAutoScaler.Name,
+			storageAutoScaler.Namespace,
+		)
+	}
+}
+
+func succeededPhaseValidation(startTime metav1.Time, completionTime metav1.Time) PhaseStatus {
+	currentTime := metav1.Now().Time
+	var zeroValuedTime metav1.Time
+	// if completion time is not set
+	// OR
+	// success happened within a 10mins time window
+	// we return success
+	if completionTime.Sub(zeroValuedTime.Time) == 0 ||
+		currentTime.Sub(completionTime.Time) < 10*time.Minute {
+		return SucceededPhaseStatus
+	}
+	// after 10 mins, we will reset the success phase
+	return ResetPhaseStatus
```

Comment:
What is the `StorageAutoScaler.Status` equivalent of this?

Reply:
StorageAutoScaler.Status.Phase

Reply:
If there is a Reset phase, then why is special handling required for metrics? It should just be able to read the phase and convert it to metric.

If tomorrow we decide that 10mins is not good enough, you'd also need to go fix the metric which seems weird. The autoscaler implementation should not be leaked into metrics.

Reply:
Correct, 

The autoscaler will keep the phase as succeeded,
Alert is designed in a way where if it see the phase is succeeded, raise a alert for next 10 mins.

Reply:
In that case, why not just stop generating the metrics for resources that are in succeeded phase for more than 10mins instead of sending a synthetic metric to reset alert? Would that not work?

Reply:
I agree with that, we can change the low-level implementation here and `not reset` it, 
@aruniiird can you make the changes?

Reply:
Done, please take a look.